sources/packages/api/src/services/restriction/student-restriction.service.ts

```diff
@@ -326,4 +327,76 @@ export class StudentRestrictionService extends RecordDataModelService<StudentRes
     studentRestriction.creator = { id: auditUserId } as User;
     return studentRestriction;
   }
+
+  /**
+   * Verify if the student has a valid SIN to apply to the particular offering.
+   * The SIN number must be a permanent one or a temporary with expiry date later
+   * then the end date of the offering.
+   * Case the SIN is not valid the student can still apply to the offering but a
+   * restriction will be created to stop funding to be disbursed till the a valid
+   * SIN is provided.
+   * @param studentId student to be validate.
+   * @param offeringId offering to be verified.
+   * @param applicationId related application. Used as additional information case
+   * a student restriction will be created.
+   * @param auditUserId user that should be considered the one that is causing the changes.
+   * @param entityManager manages the transaction where this operation must be executed.
+   */
+  async assessSINRestrictionForOfferingId(
+    studentId: number,
+    offeringId: number,
+    applicationId: number,
+    auditUserId: number,
+    entityManager: EntityManager,
+  ): Promise<void> {
+    const hasSINRestriction = await this.studentHasRestriction(
+      studentId,
+      RestrictionCode.SINF,
+    );
+    if (hasSINRestriction) {
+      // The student already has an active SIN restriction, avoid adding it again.
+      return;
+    }
+
+    const student = await entityManager
+      .getRepository(Student)
+      .createQueryBuilder("student")
+      .select([
+        "student.id",
+        "sinValidation.id",
+        "sinValidation.isValidSIN",
+        "sinValidation.temporarySIN",
+        "sinValidation.sinExpiryDate",
+      ])
+      .innerJoin("student.sinValidation", "sinValidation")
+      .where("student.id = :studentId", { studentId })
+      .getOne();
+
+    if (!student.sinValidation.temporarySIN) {
+      // SIN is not temporary so no restriction should be created.
+      return;
+    }
+    // By default assume that the temporary SIN does not have an expiry
+    // date defined and the restriction must be created.
+    let mustCreateSINException = true;
+    if (student.sinValidation.sinExpiryDate) {
+      // Temporary SIN has an expiry date and it must be validated with the end date of the offering.
+      const offering = await entityManager
```

Comment:
We can pass the offering entity model to this method instead of offering id and avoid this additional query.

Reply:
It will force the consumers (workflow setting offering and PIR complete process) to know what properties of the offering we need for this specific part. This method is selecting exactly what is needed and I would prefer that the method's signature be specific about what is needed instead of receiving an object with a lot of properties that are not used at all.
For instance, while setting the offering id from the workflow process we just have the offering id so, so this selection will be needed always.
Selecting the offering end date is only needed when the SIN is temporary, which means that for all the cases where the SIN is not temporary this query will never be executed.